storage/src/main/java/com/tencent/rss/storage/handler/impl/LocalFileClientReadHandler.java

```diff
@@ -53,29 +60,88 @@ public LocalFileClientReadHandler(
     this.shuffleServerClients = shuffleServerClients;
   }
 
-  @Override
-  public ShuffleDataResult readShuffleData(int segmentIndex) {
+  public ShuffleIndexResult readShuffleIndex() {
+    boolean readSuccessful = false;
+    ShuffleIndexResult shuffleIndexResult = null;
+    RssGetShuffleIndexRequest request = new RssGetShuffleIndexRequest(
+        appId, shuffleId, partitionId, partitionNumPerRange, partitionNum);
+
+    for (ShuffleServerClient shuffleServerClient : shuffleServerClients) {
+      try {
+        shuffleIndexResult = shuffleServerClient.getShuffleIndex(request).getShuffleIndexResult();
+        readSuccessful = true;
+        break;
+      } catch (Exception e) {
+        LOG.warn("Failed to read shuffle index with " + shuffleServerClient.getClientInfo(), e);
+      }
+    }
+
+    if (!readSuccessful) {
+      throw new RssException("Failed to read shuffle index for appId[" + appId + "], shuffleId["
+          + shuffleId + "], partitionId[" + partitionId + "]");
+    }
+
+    return shuffleIndexResult;
+  }
+
+  public ShuffleDataResult readShuffleData(ShuffleDataSegment shuffleDataSegment) {
     boolean readSuccessful = false;
     ShuffleDataResult result = null;
+    int expectedLength = shuffleDataSegment.getLength();
+
+    if (expectedLength <= 0) {
+      throw new RssException("Failed to read shuffle data for appId[" + appId + "], shuffleId["
+          + shuffleId + "], partitionId[" + partitionId + "], "
+          + "the length field in the index segment is " + expectedLength + " <= 0!");
+    }
+
     RssGetShuffleDataRequest request = new RssGetShuffleDataRequest(
-        appId, shuffleId, partitionId, partitionNumPerRange, partitionNum, readBufferSize, segmentIndex);
+        appId,shuffleId, partitionId, partitionNumPerRange, partitionNum, readBufferSize,
+        shuffleDataSegment.getOffset(), expectedLength);
+
     for (ShuffleServerClient shuffleServerClient : shuffleServerClients) {
       try {
         RssGetShuffleDataResponse response = shuffleServerClient.getShuffleData(request);
-        result = response.getShuffleDataResult();
+        result = new ShuffleDataResult(response.getShuffleData(), shuffleDataSegment.getBufferSegments());
+
+        if (result.getData().length != expectedLength) {
```

Comment:
The length is compressed data's length?

Reply:
Consider speculative execution, the length may not be expectedLength

Reply:
Yes